File: backend/app/schemas/warehouse.py
import uuid
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, ConfigDict, Field, model_validator


# ==============================================================================
# Warehouse Schemas (Phase 086)
# ==============================================================================

class WarehouseBase(BaseModel):
    code: str = Field(..., min_length=1, max_length=50, description="Unique warehouse code identifier")
    name: str = Field(..., min_length=1, max_length=200, description="Human-readable warehouse facility name")
    description: Optional[str] = Field(None, max_length=255)
    address: Optional[str] = Field(None, max_length=255)
    city: Optional[str] = Field(None, max_length=100)
    state: Optional[str] = Field(None, max_length=100)
    country: Optional[str] = Field(None, max_length=100)
    postal_code: Optional[str] = Field(None, max_length=20)
    is_active: bool = Field(default=True, description="Whether warehouse is active for operations")


class WarehouseCreate(WarehouseBase):
    pass


class WarehouseUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=255)
    address: Optional[str] = Field(None, max_length=255)
    city: Optional[str] = Field(None, max_length=100)
    state: Optional[str] = Field(None, max_length=100)
    country: Optional[str] = Field(None, max_length=100)
    postal_code: Optional[str] = Field(None, max_length=20)
    is_active: Optional[bool] = None


class WarehouseResponse(WarehouseBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    company_id: uuid.UUID
    created_at: datetime
    updated_at: datetime
    total_stock_items: int = 0
    total_physical_stock: int = 0
    total_reserved_stock: int = 0
    total_atp: int = 0


class WarehouseListResponse(BaseModel):
    items: List[WarehouseResponse]
    total: int
    page: int
    size: int
    pages: int


# ==============================================================================
# Warehouse Stock Schemas (Phase 087)
# ==============================================================================

class WarehouseStockCreate(BaseModel):
    product_id: uuid.UUID
    quantity: int = Field(default=0, ge=0, description="Physical stock quantity on hand")
    reserved_quantity: int = Field(default=0, ge=0, description="Stock allocated/reserved")

    @model_validator(mode="after")
    def validate_reservation_not_exceed_quantity(self):
        if self.reserved_quantity > self.quantity:
            raise ValueError("reserved_quantity cannot exceed physical stock quantity")
        return self


class WarehouseStockUpdate(BaseModel):
    quantity: int = Field(..., ge=0, description="New physical stock quantity")


class WarehouseStockResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    warehouse_id: uuid.UUID
    product_id: uuid.UUID
    quantity: int
    reserved_quantity: int
    available_to_promise: int
    is_available: bool
    product_sku: Optional[str] = None
    product_name: Optional[str] = None
    product_unit: Optional[str] = None
    category_name: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class WarehouseStockListResponse(BaseModel):
    warehouse_id: uuid.UUID
    warehouse_code: str
    warehouse_name: str
    items: List[WarehouseStockResponse]
    total: int
    total_physical: int
    total_reserved: int
    total_atp: int


# ==============================================================================
# Stock Availability Schemas (Phase 088)
# ==============================================================================

class StockAvailabilityResponse(BaseModel):
    product_id: uuid.UUID
    product_name: str
    product_sku: str
    warehouse_id: uuid.UUID
    warehouse_name: str
    warehouse_code: str
    stock_quantity: int
    reserved_quantity: int
    available_quantity: int
    is_available: bool


# ==============================================================================
# Reserved Stock Schemas (Phase 089)
# ==============================================================================

class StockReserveRequest(BaseModel):
    quantity: int = Field(..., gt=0, description="Quantity to reserve (must be strictly positive)")


class StockReleaseRequest(BaseModel):
    quantity: int = Field(..., gt=0, description="Quantity to release from reservation (must be strictly positive)")


# ==============================================================================
# Available-to-Promise (ATP) Schemas (Phase 090)
# ==============================================================================

class ATPResponse(BaseModel):
    product_id: uuid.UUID
    warehouse_id: uuid.UUID
    physical_stock: int
    reserved_stock: int
    available_to_promise: int
    is_available: bool

File: frontend/src/types/warehouse.ts
/**
 * TypeScript Interfaces for Warehouse & Inventory (G18: Phases 086–090).
 */

export interface Warehouse {
  id: string;
  company_id: string;
  code: string;
  name: string;
  description: string | null;
  address: string | null;
  city: string | null;
  state: string | null;
  country: string | null;
  postal_code: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  total_stock_items: number;
  total_physical_stock: number;
  total_reserved_stock: number;
  total_atp: number;
}

export interface WarehouseCreateInput {
  code: string;
  name: string;
  description?: string | null;
  address?: string | null;
  city?: string | null;
  state?: string | null;
  country?: string | null;
  postal_code?: string | null;
  is_active?: boolean;
}

export interface WarehouseUpdateInput {
  name?: string;
  description?: string | null;
  address?: string | null;
  city?: string | null;
  state?: string | null;
  country?: string | null;
  postal_code?: string | null;
  is_active?: boolean;
}

export interface WarehouseListResponse {
  items: Warehouse[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export interface WarehouseStock {
  id: string;
  warehouse_id: string;
  product_id: string;
  quantity: number;
  reserved_quantity: number;
  available_to_promise: number;
  is_available: boolean;
  product_sku: string | null;
  product_name: string | null;
  product_unit: string | null;
  category_name: string | null;
  created_at: string;
  updated_at: string;
}

export interface WarehouseStockListResponse {
  warehouse_id: string;
  warehouse_code: string;
  warehouse_name: string;
  items: WarehouseStock[];
  total: number;
  total_physical: number;
  total_reserved: number;
  total_atp: number;
}

export interface WarehouseStockCreateInput {
  product_id: string;
  quantity: number;
  reserved_quantity?: number;
}

export interface WarehouseStockUpdateInput {
  quantity: number;
}

export interface StockReserveReleaseInput {
  quantity: number;
}

export interface StockAvailability {
  product_id: string;
  product_name: string;
  product_sku: string;
  warehouse_id: string;
  warehouse_name: string;
  warehouse_code: string;
  stock_quantity: number;
  reserved_quantity: number;
  available_quantity: number;
  is_available: boolean;
}

export interface ATPData {
  product_id: string;
  warehouse_id: string;
  physical_stock: number;
  reserved_stock: number;
  available_to_promise: number;
  is_available: boolean;
}
